### ISSUE TITLE
Introduce better support for accessing fields

### ISSUE THREAD

**Issue by britter** [junit-team/junit5#1611]
## Feature Request

_Add APIs for accessing fields of test classes and instances._

When authoring an extension one often needs to access fields of a test class or instance fields of a test instance. Usually one is interested in fields of a certain type possibly annotated with a certain method. There is currently no API to do this in JUnit so one has to fall back to raw reflection code.

I see two ways to implement this:

1. Add more methods to `ReflectionSupport` - this will work but I don't think we should put more and more functionality into these helper classes

```java
/**
 * Get the values of the static fields annotated with the given annotation from the given class.
 */
List<Object> getAnnotatedFields(Class<?>, Class<? extends Annotation>)

/**
 * Get the values of the static fields annotated with the given annotation of the given
 * type from the given class.
 */
<T> List<T> ReflectionSupport.getAnnotatedFieldsOfType(Class<?>, Class<? extends Annotation>, Class<T> type)

/**
 * Get the values of the instance fields annotated with the given annotation from the given object.
 */
List<Object> ReflectionSupport.getAnnotatedFields(Object, Class<? extends Annotation>)

/**
 * Get the values of the instance fields annotated with the given annotation of the given
 * type from the given object.
 */
<T> List<T> ReflectionSupport.getAnnotatedFieldsOfType(Object, Class<? extends Annotation>, Class<T> type)

```

2. Follow a more object-oriented approach and create specialized objects that help with this:

```java
/**
 * A wrapper around the test class being executed.
 */
public interface TestClass {

    List<Object> getAnnotatedFieldValues(Class<? extends Annotation>)

    <T>  List<T> getAnnotatedFieldValues(Class<? extends Annotation>, Class<T>)
}

/**
 * A wrapper around the test instance being executed.
 */
public interface TestInstance {

    List<Object> getAnnotatedFieldValues(Class<? extends Annotation>)

    <T>  List<T> getAnnotatedFieldValues(Class<? extends Annotation>, Class<T>)
} 
```

The specialized objects could be accessed from the `ExtensionContext`. Methods like `getRequiredTestClass` et al. could be deprecated.

## Deliverables

- [x] Introduce API for easier access of fields inside of test classes and test instances.
- [x] Document in User Guide.
   - ❌ deemed unnecessary since the new features are documented in Javadoc.
- [x] Document in Release Notes.

**Comment by britter** [comment]
The `ExtensionContext` should give access to these wrapper classes (there should probably also be a `TestMethod` object). They could all provide getters to the wrapped objects. Evolving the API will be a challenge here. I see several ways:

1. find a good name for the accessors to the wrapped objects. As you pointed out the obvious names are already taken. So one would have to find a different name.
2. Deprecate `ExtensionContext` and define new interfaces. The current design does not follow [LSP](https://en.wikipedia.org/wiki/Liskov_substitution_principle): You can't swap out a `TestMethodExtensionContext` with a `ClassExtensionContext` without changing the correctness of the program. This is because `ExtensionContext` is a one fits all interface. But having `getTestMethod()` on the `ExtensionContext` that is passed to `beforeAll` does not make sense. So I would recommend to create specialized ExtensionContext on those could have the new methods described above. The same strange design can be observed in the `TestInstancePostProcessor` hook: it get's passed the test instance that has just been created, but calling `getTestInstance` on the provided `ExtensionContext` will yield `Optional.empty`.

**Comment by marcphilipp** [comment]
Let's try to objectively assess advantages and disadvantages of the two approaches. Static helper methods are certainly less discoverable. However, they can be used for more than just test classes, e.g. for finding fields on classes in general. What else?

**Comment by zoltanersek** [comment]
Other advantages for static helper methods: 

There would be no change to ExtensionContext, which would not break existing clients when they try to upgrade in the future. 

There would not be other classes instantiated.

**Comment by britter** [comment]
The static helpers can be implemented right away. They could even be used in the wrappers should they be implemented at some point in time.

**Comment by sbrannen** [comment]
Regarding naming conventions for methods and `ReflectionSupport` vs. `AnnotationSupport`, please see https://github.com/junit-team/junit5/pull/1617#issuecomment-427583980.

**Comment by sbrannen** [comment]
> When authoring an extension one often needs to access fields of a test class or instance fields of a test instance. Usually one is interested in fields of a certain type possibly annotated with a certain method. There is currently no API to do this in JUnit so one has to fall back to raw reflection code.

FYI: there actually is (some) existing support: `AnnotationSupport.findPublicAnnotatedFields(...)`, `AnnotationUtils.findAnnotatedFields(...)`, and `ReflectionUtils.readFieldValue(...)`, although the latter two are technically not _public_. 😉

**Comment by sbrannen** [comment]
> Deprecate `ExtensionContext` and define new interfaces. 

My gut feeling is that I would object to any such deprecation.

**Comment by sbrannen** [comment]
> The static helpers can be implemented right away. They could even be used in the wrappers should they be implemented at some point in time.

Implementing the utility methods in `AnnotationSupport` is a no-brainer IMHO, especially for methods that already exist in `AnnotationUtils` and `ReflectionUtils`.

**Comment by britter** [comment]
> > Deprecate `ExtensionContext` and define new interfaces.
> 
> My gut feeling is that I would object to any such deprecation.

I understand that deprecating an API that has just been released feels wrong. But I don't think that the XxxUtils/Support approach is a good idea. It's hard to discover for users and Utils always end up as a dumping ground for everything remotely related.

**Comment by britter** [comment]
May proposal for an improved extension API can be found here: https://github.com/britter/improved-extension-api

**Comment by npathai** [comment]
Is someone working on this issue? Else I would like to contribute.

**Comment by marcphilipp** [comment]
@npathai Thanks for asking! There's already a PR for it: #1617.

**Comment by sbrannen** [comment]
FYI: the current work on this issue can be seen in the following branch -- which builds on the work from PR #1617.

https://github.com/junit-team/junit5/compare/issues/1611-field-support-utils

**Comment by sbrannen** [comment]
**Update**:

New API is visible here: https://github.com/junit-team/junit5/compare/issues/1611-field-support-utils

Documentation in release notes and user guide is forthcoming.

**Comment by sbrannen** [comment]
Closing this issue since the features targeted for 5.4 have been merged into `master` in commit 3e3a53b8f5d27f3696f9813aa406894102b19a17.

If anyone wishes to propose a "more object-oriented approach" as mentioned by @britter, please do so in a new issue.

Thanks again to @zoltanersek for the initial work that went into the new feature set.
